pennylane/workflow/set_shots.py

```diff
@@ -16,75 +16,148 @@
-    """Transform used to set or update a circuit's shots.
+    *, shots: ShotsLike
+) -> Callable[[QNode], QNode]: ...
+
+
+def set_shots(*args, shots: ShotsLike = _SHOTS_NOT_PROVIDED):
+    """Transform used to set or update a circuit's shots."""
+    # Keyword-only case: @set_shots(shots=500) or @set_shots(shots=None)
+    if len(args) == 0 and shots is not _SHOTS_NOT_PROVIDED:
+        return _create_keyword_decorator(shots)
+
+    if len(args) == 1 and shots is not _SHOTS_NOT_PROVIDED:
+        # Direct application: set_shots(qnode, shots=500) or set_shots(qnode, shots=None)
+        return _apply_shots_to_qnode(args[0], shots)
+    if len(args) == 1 and shots is _SHOTS_NOT_PROVIDED:
+        # Positional decorator: @set_shots(500) or @set_shots(None)
+        return _set_shots_dispatch(args[0])
+
+    raise ValueError(f"Invalid arguments to set_shots: {args=}, {shots=}")
+
+
+def _set_shots_dispatch(shots_value: ShotsLike) -> Callable[[QNode], QNode]:
+    """Default case: @set_shots(500) - positional shots value"""
+
+    def positional_decorator(qnode_func: QNode) -> QNode:
+        return qnode_func.update_shots(shots_value)
+
+    return positional_decorator
 
-    Args:
-        qnode (QNode): The QNode to transform. If not provided, `set_shots` can be used as a decorator directly.
-        shots (None or int or Sequence[int] or Sequence[tuple[int, int]] or pennylane.shots.Shots): The
-            number of shots (or a shots vector) that the transformed circuit will execute.
 
-    Returns:
-        QNode or callable: The transformed QNode with updated shots, or a wrapper function
-        if qnode is not provided.
+def _apply_shots_to_qnode(qnode: QNode, shots: ShotsLike) -> QNode:
+    """Handle direct application to a QNode: set_shots(qnode, shots=500)"""
+    if not isinstance(qnode, QNode):
+        raise ValueError("set_shots can only be applied to QNodes")
+    return qnode.update_shots(shots)
 
-    There are three ways to specify shot values (see :func:`qml.measurements.Shots <pennylane.measurements.Shots>` for more details):
 
-    * The value ``None``: analytic mode, no shots
-    * A positive integer: a fixed number of shots
-    * A sequence consisting of either positive integers or a tuple-pair of positive integers of the form ``(shots, copies)``
+def _create_keyword_decorator(shots: ShotsLike) -> Callable[[QNode], QNode]:
+    """Handle keyword-only decorator: @set_shots(shots=500)"""
 
-    **Examples**
+    def decorator(qnode_func: QNode) -> QNode:
+        return qnode_func.update_shots(shots)
 
-    Set the number of shots as a decorator:
+    return decorator
 
-    .. code-block:: python
 
-        @qml.set_shots(shots=2)
-        @qml.qnode(qml.device("default.qubit", wires=1))
-        def circuit():
-            qml.RX(1.23, wires=0)
-            return qml.sample(qml.Z(0))
+# Add comprehensive docstring and examples
+set_shots.__doc__ = """Transform used to set or update a circuit's shots.
```

Comment:
Why is the docstring here?

Reply:
lol I used to implement it in a very dynamic multidispatching way so I separate the doc here; but now it should be fine to go back to the main set_shots

Reply:
https://github.com/PennyLaneAI/pennylane/pull/7919/commits/715caa89b6f86f726bcf23a989974cc2fe2743e4